src/video.cpp

```diff
@@ -791,24 +791,46 @@ namespace video {
     }
     display_wp = disp;
 
-    std::vector<std::shared_ptr<platf::img_t>> imgs(12);
-    auto round_robin = round_robin_util::make_round_robin<std::shared_ptr<platf::img_t>>(std::begin(imgs), std::end(imgs));
-
-    for (auto &img : imgs) {
-      img = disp->alloc_img();
-      if (!img) {
-        BOOST_LOG(error) << "Couldn't initialize an image"sv;
-        return;
+    constexpr auto capture_buffer_normal_size = 1;
+    constexpr auto capture_buffer_size = 12;
+    std::list<std::shared_ptr<platf::img_t>> imgs(capture_buffer_size);
+
+    auto pull_free_image_callback = [&](std::shared_ptr<platf::img_t> &img_out) -> bool {
+      img_out.reset();
+      for (auto it = imgs.begin(); it != imgs.end(); it++) {
+        // pick first unallocated or unused
+        if (!*it || it->use_count() == 1) {
+          if (!*it) {
+            // allocate if unallocated
+            *it = disp->alloc_img();
+          }
+          img_out = *it;
+          if (it != imgs.begin()) {
+            // move freshly allocated or unused img to the front of the list to prioritize its reusal
+            imgs.erase(it);
+            imgs.push_front(img_out);
+          }
+          break;
+        }
       }
-    }
+      if (img_out) {
```

Comment:
If we don't manage to find a free image, we should sleep and try again until we do get one. There can be transient conditions that might hang the encoder temporarily, but we should just wait for one to free up (like the current code does) rather than returning an error.

Making this function always succeed will also allow you to eliminate the error handling around this new callback which will simplify the code.

Reply:
I can do that. With interrupt signal or variable captured by the callback. Giving control over sleeps to callers has its own pros, but going with as little changes unrelated to vram usage is the right call.

Reply:
If I find the signal :(
Variable won't cut it because everything is likely to happen in single capture thread.

Reply:
I guess I can check for `capture_ctx_queue->running()` https://github.com/ns6089/Sunshine/blob/remove_round_robin/src/video.cpp#L830 and hope it's enough to prevent potential deadlocks.

Reply:
Like this https://github.com/LizardByte/Sunshine/pull/1098/commits/4e5dfc3de217087891f80f8aef4f339c44244864